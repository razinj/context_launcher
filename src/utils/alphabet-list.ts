import { AppDetails } from '../models/app-details'
import { AppLetterIndex } from '../models/list-letter-index'

type LetterMap = {
  [key: string]: number
}

const alphabet = [
  'A',
  'B',
  'C',
  'D',
  'E',
  'F',
  'G',
  'H',
  'I',
  'J',
  'K',
  'L',
  'M',
  'N',
  'O',
  'P',
  'Q',
  'R',
  'S',
  'T',
  'U',
  'V',
  'W',
  'X',
  'Y',
  'Z',
]

/**
 * @returns Indexed alphabet letters in an object where the key is the alphabet and the value is the index
 */
export const getLettersMap = (): LetterMap => {
  const letterMap: LetterMap = {}

  alphabet.forEach((letter: string, index: number) => (letterMap[letter] = index + 1))

  return letterMap
}

/**
 * @param value Any string value
 * @returns A letter if the first chareceter of the value is an alphabet, otherwise it returns a '#'
 */
export const getFirstLetter = (value: string): string => {
  const firstChar = value.trim().substring(0, 1).toUpperCase()
  const isValidLetter = getLettersMap()[firstChar]

  if (isValidLetter) return firstChar

  return '#'
}

export const getAppsLetterIndex = (apps: AppDetails[]): AppLetterIndex[] => {
  const appsLetterIndex: AppLetterIndex[] = []
  const treatedLetters: string[] = []

  apps.forEach(({ name }: AppDetails, index: number) => {
    const letter = getFirstLetter(name)

    if (treatedLetters.includes(letter)) return

    appsLetterIndex.push({ letter, index })
    treatedLetters.push(letter)
  })

  return appsLetterIndex
}
