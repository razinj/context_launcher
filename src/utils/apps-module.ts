// Native modules
import AppsModule from '../native-modules/AppsModule'

export const launchApp = (packageName: string) => AppsModule.launchApplication(packageName)

export const showAppDetails = (packageName: string) => AppsModule.showApplicationDetails(packageName)

export const requestAppUninstall = (packageName: string) => AppsModule.requestApplicationUninstall(packageName)
