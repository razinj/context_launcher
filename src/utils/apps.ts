// Models
import { AppDetails, AppDetailsWithIcon } from '../models/app-details'
import { PinnedApp } from '../models/pinned-app'

export const getAppsByLabel = (apps: AppDetails[], query: string): AppDetails[] => {
  return apps.filter((app: AppDetails) => app.label.match(new RegExp(query, 'gi')))
}

export const getPinnedAppByName = (pinnedApps: PinnedApp[], appDetails: AppDetailsWithIcon): PinnedApp => {
  const pinnedApp = pinnedApps.find(({ name }: PinnedApp) => name === appDetails.name)

  if (pinnedApp) return pinnedApp

  return { ...appDetails, isTemporary: false, isPermanent: false }
}
