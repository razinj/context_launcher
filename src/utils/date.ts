export const getTimeFromDate = (stringDate?: string): string => {
  if (!stringDate) return 'unset'

  const date = new Date(stringDate)

  const singleDigitHours = `${date.getHours()}`.length == 1
  const singleDigitMinutes = `${date.getMinutes()}`.length == 1

  const hours = singleDigitHours ? `0${date.getHours()}` : date.getHours()
  const minutes = singleDigitMinutes ? `0${date.getMinutes()}` : date.getMinutes()

  return `${hours}:${minutes}`
}

export const getDateFromStringWithCurrentDateValue = (stringDate?: string): Date | undefined => {
  if (!stringDate) return

  const currentDate = new Date()
  const fullDate = new Date(stringDate)
  return new Date(
    currentDate.getFullYear(),
    currentDate.getMonth(),
    currentDate.getDate(),
    fullDate.getHours(),
    fullDate.getMinutes(),
    0,
    0
  )
}

export const stripDateFromSeconds = (date: Date): Date => {
  return new Date(date.getFullYear(), date.getMonth(), date.getDate(), date.getHours(), date.getMinutes(), 0, 0)
}
