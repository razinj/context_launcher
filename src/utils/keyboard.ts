import { Keyboard } from 'react-native'

export const dismissKeyboard = () => Keyboard.dismiss()
