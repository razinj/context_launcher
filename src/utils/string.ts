export const truncateString = (value: string | undefined, maxLength: number): string | undefined => {
  return value && value.length > maxLength ? `${value.substring(0, maxLength)}...` : value
}
