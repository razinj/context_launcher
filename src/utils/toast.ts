import { ToastAndroid } from 'react-native'

export const displayToast = (message: string, duration: number = ToastAndroid.SHORT) => {
  ToastAndroid.show(message, duration)
}
