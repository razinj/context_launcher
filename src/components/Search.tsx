// React
import React, { useContext, useEffect } from 'react'
// React Native
import { Pressable, PressableAndroidRippleConfig, StyleSheet, TextInput, View } from 'react-native'
// Redux
import { useDispatch, useSelector } from 'react-redux'
import {
  resetAppsSearchState,
  selectAppsSearchQuery,
  setAppsSearchQuery,
  setAppsSearchResult,
} from '../slices/appsSearch'
import { selectAppsListMemoized } from '../slices/appsList'
// Contexts
import SearchContext from '../contexts/SearchContext'
import GlobalContext from '../contexts/GlobalContext'
// Icon
import Icon from 'react-native-vector-icons/MaterialIcons'
// Utils
import { dismissKeyboard } from '../utils/keyboard'
// Constants
import { PRIMARY_HEX_COLOR, SECONDARY_HEX_COLOR } from '../constants'
// Models
import { AppDetails } from '../models/app-details'

const clearIconRippleConfig: PressableAndroidRippleConfig = {
  borderless: true,
  foreground: true,
  color: '#ccc',
  radius: 15,
}

const Search = () => {
  const dispatch = useDispatch()
  const apps = useSelector(selectAppsListMemoized)
  const searchQuery = useSelector(selectAppsSearchQuery)
  const { searchInputRef, invalidCharacters, setInvalidCharacters } = useContext(SearchContext)
  const { hideAllApps, displayAllApps } = useContext(GlobalContext)

  const onQueryChange = (query: string) => {
    const trimmedQuery = query.trim().replace(/\./g, '\\.')

    // Accept only ASCII based characters
    if (!trimmedQuery.match(/[A-z\s\d]/gi)) {
      setInvalidCharacters(true)
      dispatch(setAppsSearchQuery(trimmedQuery))
      return
    } else if (trimmedQuery.length === 0) {
      dispatch(resetAppsSearchState())
      return
    }

    // Reset invalid characters when it's valid (passes the above check)
    if (invalidCharacters) setInvalidCharacters(false)

    hideAllApps()

    dispatch(setAppsSearchResult(apps.filter((app: AppDetails) => app.label.match(new RegExp(trimmedQuery, 'gi')))))
    dispatch(setAppsSearchQuery(trimmedQuery))
  }

  const clearInputAndSearchState = () => {
    searchInputRef?.current?.clear()
    dispatch(resetAppsSearchState())
  }

  const onSubmitEditing = () => {
    dismissKeyboard()
  }

  useEffect(() => {
    if (!searchQuery) searchInputRef?.current?.clear()
  }, [searchQuery])

  return (
    <View style={styles.wrapper}>
      <TextInput
        ref={searchInputRef}
        style={styles.searchInput}
        placeholder='Search'
        placeholderTextColor={displayAllApps ? PRIMARY_HEX_COLOR : SECONDARY_HEX_COLOR}
        returnKeyType='search'
        autoCapitalize='words'
        onChangeText={onQueryChange}
        onSubmitEditing={onSubmitEditing}
      />
      {searchQuery && searchQuery?.length > 0 && (
        <Pressable
          style={styles.clearIconWrapper}
          onPress={clearInputAndSearchState}
          android_disableSound={true}
          android_ripple={clearIconRippleConfig}>
          <Icon name='clear' size={25} color='#808080' />
        </Pressable>
      )}
    </View>
  )
}

const styles = StyleSheet.create({
  wrapper: {
    display: 'flex',
    alignItems: 'center',
    flexDirection: 'row',
    paddingHorizontal: 5,
    flex: 1,
  },
  searchInput: {
    flex: 1,
    color: '#000',
    fontSize: 16,
  },
  clearIconWrapper: {
    padding: 5,
  },
})

export default Search
