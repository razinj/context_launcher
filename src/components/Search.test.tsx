import { fireEvent } from '@testing-library/react-native'
import { initialStoreState } from '../../utils/test/data'
import { renderWithProviderAndContexts, renderWithProvider } from '../../utils/test/utils'
import Search from './Search'

// TODO: Cover all possible cases for the search functionality
describe('<Search /> Tests', () => {
  it('should render correctly and match snapshot', () => {
    const { toJSON, getByTestId, queryByTestId } = renderWithProvider(<Search />)

    expect(toJSON()).toMatchSnapshot()
    expect(getByTestId('search-input')).toBeDefined()
    expect(queryByTestId('search-input-clear-button')).toBeNull()
  })

  it('should render correctly and match snapshot with clear button', () => {
    const customInitialState = {
      ...initialStoreState,
      appsSearch: {
        ...initialStoreState.appsSearch,
        query: 'a-search-query',
      },
    }

    const { getByTestId, toJSON } = renderWithProvider(<Search />, { preloadedState: customInitialState })

    expect(toJSON()).toMatchSnapshot()
    expect(getByTestId('search-input')).toBeDefined()
    expect(getByTestId('search-input-clear-button')).toBeDefined()
  })

  it('should update query when text changes and clear query when clear button is pressed', () => {
    const customInitialState = {
      ...initialStoreState,
      appsSearch: {
        ...initialStoreState.appsSearch,
        query: undefined,
      },
    }

    const { store, getByTestId } = renderWithProviderAndContexts(<Search />, { preloadedState: customInitialState })

    const searchInput = getByTestId('search-input')

    expect(searchInput).toBeDefined()

    fireEvent.changeText(searchInput, 'a-search-query')

    expect(store.getState().appsSearch.query).toBe('a-search-query')

    const searchInputClearButton = getByTestId('search-input-clear-button')

    expect(searchInputClearButton).toBeDefined()

    fireEvent.press(searchInputClearButton)

    expect(store.getState().appsSearch.query).toBeUndefined()
  })

  it('should set correct result list when using correct query and reset search values when clear button is pressed', () => {
    const customInitialState = {
      ...initialStoreState,
      appsList: {
        list: [
          {
            name: 'com.google.chrome',
            label: 'Chrome',
          },
          {
            name: 'com.google.maps',
            label: 'Maps',
          },
        ],
      },
      appsSearch: {
        query: undefined,
        result: [],
      },
    }

    const { store, getByTestId } = renderWithProviderAndContexts(<Search />, { preloadedState: customInitialState })

    const searchInput = getByTestId('search-input')

    expect(searchInput).toBeDefined()

    fireEvent.changeText(searchInput, '  chr  ') // Whitespace to test trimming

    let currentState = store.getState()

    expect(currentState.appsSearch.query).toBe('chr')
    expect(currentState.appsSearch.result).toEqual([
      {
        name: 'com.google.chrome',
        label: 'Chrome',
      },
    ])

    const searchInputClearButton = getByTestId('search-input-clear-button')

    expect(searchInputClearButton).toBeDefined()

    fireEvent.press(searchInputClearButton)

    currentState = { ...store.getState() }

    expect(currentState.appsSearch.query).toBeUndefined()
    expect(currentState.appsSearch.result).toEqual([])
  })

  it('should set correct result list when using wrong query and reset search values when clear button is pressed', () => {
    const customInitialState = {
      ...initialStoreState,
      appsList: {
        list: [
          {
            name: 'com.google.chrome',
            label: 'Chrome',
          },
        ],
      },
      appsSearch: {
        query: undefined,
        result: [],
      },
    }

    const { store, getByTestId } = renderWithProviderAndContexts(<Search />, { preloadedState: customInitialState })

    const searchInput = getByTestId('search-input')

    expect(searchInput).toBeDefined()

    fireEvent.changeText(searchInput, 'youtube')

    let currentState = store.getState()

    expect(currentState.appsSearch.query).toBe('youtube')
    expect(currentState.appsSearch.result).toEqual([])

    const searchInputClearButton = getByTestId('search-input-clear-button')

    expect(searchInputClearButton).toBeDefined()

    fireEvent.press(searchInputClearButton)

    currentState = { ...store.getState() }

    expect(currentState.appsSearch.query).toBeUndefined()
    expect(currentState.appsSearch.result).toEqual([])
  })
})
