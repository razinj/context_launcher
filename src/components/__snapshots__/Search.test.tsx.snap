// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Search /> Tests should render correctly and match snapshot 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "display": "flex",
      "flex": 1,
      "flexDirection": "row",
      "paddingHorizontal": 5,
    }
  }
>
  <TextInput
    autoCapitalize="words"
    onChangeText={[Function]}
    onSubmitEditing={[Function]}
    placeholder="Search"
    placeholderTextColor="#A0C2AD"
    returnKeyType="search"
    style={
      {
        "color": "#000",
        "flex": 1,
        "fontSize": 16,
      }
    }
    testID="search-input"
  />
</View>
`;

exports[`<Search /> Tests should render correctly and match snapshot with clear button 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "display": "flex",
      "flex": 1,
      "flexDirection": "row",
      "paddingHorizontal": 5,
    }
  }
>
  <TextInput
    autoCapitalize="words"
    onChangeText={[Function]}
    onSubmitEditing={[Function]}
    placeholder="Search"
    placeholderTextColor="#A0C2AD"
    returnKeyType="search"
    style={
      {
        "color": "#000",
        "flex": 1,
        "fontSize": 16,
      }
    }
    testID="search-input"
  />
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "padding": 5,
      }
    }
    testID="search-input-clear-button"
  >
    <Text
      allowFontScaling={false}
      selectable={false}
      style={
        [
          {
            "color": "#808080",
            "fontSize": 25,
          },
          undefined,
          {
            "fontFamily": "Material Icons",
            "fontStyle": "normal",
            "fontWeight": "normal",
          },
          {},
        ]
      }
    >
      î…Œ
    </Text>
  </View>
</View>
`;
